### PR TITLE
[SPARK-39958][SQL] Add warning log when unable to load custom metric object

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

Currently when `SQLAppStatusListener` cannot load custom metric object, it silently ignores the error. We better provide a warning log for users.

### Why are the changes needed?

Improving user experience.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing tests

### CODE REVIEW

**Comment by viirya** [comment]
cc @dongjoon-hyun

**Comment by pan3793** [comment]
Previous discussion: https://github.com/apache/spark/pull/31451#discussion_r935521681

I think we don't need to construct the instance using reflection, how about removing constructor checks?

**Comment by cloud-fan** [comment]
> I think we don't need to construct the instance using reflection

Then how can we construct the instance?

**Comment by viirya** [comment]
Just about to ask same question.

**Comment by pan3793** [comment]
Oh, I see. `SQLMetrics.createV2CustomMetric` only store the class name instead of the instance itself, then constructing the instance by reflection is necessary.

Previous I misunderstand that instances constructed by `Scan#supportedCustomMetrics` will be passed and reused.

Could you please enhance the comment of `CustomMetric` to mention the constructor requirements as well?

**Comment by viirya** [comment]
> Could you please enhance the comment of CustomMetric to mention the constructor requirements as well?

Okay.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala [hunk 1]
```diff
@@ -220,7 +220,10 @@ class SQLAppStatusListener(
             metricAggregationMap.put(className, method)
             method
           } catch {
-            case NonFatal(_) =>
+            case NonFatal(e) =>
+              logWarning(s"Unable to load custom metric object for class `$className`. " +
+                "Please make sure that the custom metric class is in the classpath and " +
+                "it has 0-arg constructor", e)
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala:226]: nit. `.` at the end of the sentence?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala (hunk 1 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala]: Added. Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala [hunk 2]
```diff
@@ -220,7 +220,10 @@ class SQLAppStatusListener(
             metricAggregationMap.put(className, method)
             method
           } catch {
-            case NonFatal(_) =>
+            case NonFatal(e) =>
+              logWarning(s"Unable to load custom metric object for class `$className`. " +
+                "Please make sure that the custom metric class is in the classpath and " +
+                "it has 0-arg constructor.", e)
```
Comment (pan3793) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala:226]: `Utils#loadExtensions` also allows "single-argument constructor that accepts SparkConf", do we need to mention that?

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala:226]: I don't think we should. Actually we don't need the implementing class to implement it.

**Review by pan3793 — APPROVED** [review]
Thanks, LGTM

**Comment by viirya** [comment]
The known and unrelated sparkr failure:

```
Status: 1 WARNING, 2 NOTEs
See
  ‘/__w/spark-1/spark-1/R/SparkR.Rcheck/00check.log’
for details.


+ popd
Had CRAN check errors; see logs.
[error] running /__w/spark-1/spark-1/R/run-tests.sh ; received return code 255
```


Unrelated Python linter:

```
Traceback (most recent call last):
  File "/usr/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.9/dist-packages/black/__main__.py", line 3, in <module>
    patched_main()
  File "/usr/local/lib/python3.9/dist-packages/black/__init__.py", line 1372, in patched_main
    patch_click()
  File "/usr/local/lib/python3.9/dist-packages/black/__init__.py", line 1358, in patch_click
    from click import _unicodefun
ImportError: cannot import name '_unicodefun' from 'click' (/usr/local/lib/python3.9/dist-packages/click/__init__.py)
Please run 'dev/reformat-python' script.
```

**Comment by viirya** [comment]
Merged, thanks!

**Pull request closed without merging**